### PR TITLE
Trim hls

### PR DESCRIPTION
**Pull request by krystalmejia24**
Refs issue #37 

There are some assumptions written in the issue that I'd like to clarify. May need to make some changes to propeller for accessing that bakery URL. CC @allisonnicoledeal 

When you request a propeller stream via bakery in the form of 

`https://bakery.dev.cbsivideo.com/t(1583971200,1584057600)/propeller/orgID/channelID.m3u8`

Bakery will fetch the Propeller stream and rewrite the variant URLs to point to bakery in the form of 

`https://bakery.dev.cbsivideo.com/t(1583971200,1584057600)/<base64-encoded-original-media-manifest-url>.m3u8`

Currently, there is a dependency on PDT for every segments. Since propeller is adding PDT to each segment we will be fine with trimming any streams with a Propeller origin.

### CODE REVIEW

File: pkg/filters/hls.go [hunk 1]
```diff
@@ -174,6 +186,18 @@ func (h *HLSFilter) normalizeVariant(v *m3u8.Variant, absolute url.URL) (*m3u8.V
 	return v, nil
 }
 
+func (h *HLSFilter) normalizeTrimmedVariant(filters *parsers.MediaFilters, uri string) (string, error) {
+	encoded := base64.StdEncoding.EncodeToString([]byte(uri))
+	start := filters.Trim.Start
+	end := filters.Trim.End
+	u, err := url.Parse(uri)
```
Comment (zsiec) [inline pkg/filters/hls.go]: Will this work for master manifests with relative paths to media manifests?

File: pkg/filters/hls.go [hunk 2]
```diff
@@ -174,6 +186,18 @@ func (h *HLSFilter) normalizeVariant(v *m3u8.Variant, absolute url.URL) (*m3u8.V
...
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%v://%v/t(%v,%v)/%v.m3u8", u.Scheme, h.config.Hostname, start, end, encoded), nil
```
Comment (zsiec) [inline pkg/filters/hls.go]: Curious if this works for master manifests with relative paths to media manifests?

File: pkg/filters/hls.go [hunk 3]
```diff
@@ -200,22 +224,56 @@ func isRelative(urlStr string) (bool, error) {
...
 
-// Returns true if given codec is an audio codec (mp4a, ec-3, or ac-3)
-func isAudioCodec(codec string) bool {
-	return (ValidCodecs(codec, aacCodec) ||
-		ValidCodecs(codec, ec3Codec) ||
-		ValidCodecs(codec, ac3Codec))
+// FilterRenditionManifest will be responsible for filtering the manifest
+// according  to the MediaFilters
+func (h *HLSFilter) filterRenditionManifest(filters *parsers.MediaFilters, m *m3u8.MediaPlaylist) (string, error) {
+	filteredPlaylist, _ := m3u8.NewMediaPlaylist(uint(len(m.Segments)), uint(len(m.Segments)))
```
Comment (zsiec) [inline pkg/filters/hls.go]: I think we should handle this error by early returning it to the caller.

File: pkg/filters/hls.go [hunk 4]
```diff
@@ -200,22 +224,56 @@ func isRelative(urlStr string) (bool, error) {
...
+	seqID := 0
```
Comment (zsiec) [inline pkg/filters/hls.go]: When declaring a variable initialized with its default value, you could use `var seqID int` instead of the `:=` assignment.

File: pkg/filters/hls.go [hunk 5]
```diff
@@ -200,22 +224,56 @@ func isRelative(urlStr string) (bool, error) {
...
+
+	for _, segment := range m.Segments {
+		if segment != nil {
```
Comment (zsiec) [inline pkg/filters/hls.go]: I find that deep nesting can get hard to follow, you can remove one level by doing:

```go
if segment == nil {
    continue
}

File: pkg/filters/hls.go [hunk 6]
```diff
@@ -200,22 +224,56 @@ func isRelative(urlStr string) (bool, error) {
...
+					return "", fmt.Errorf("trimming segments: %w", err)
+				}
+
+				if isRelativeURL {
+					absolute, err := getAbsoluteURLString(h.manifestURL)
+					if err != nil {
+						return "", fmt.Errorf("fetching absolute url: %w", err)
+					}
+
+					segment.URI = absolute + segment.URI
```
Comment (zsiec) [inline pkg/filters/hls.go]: You may want to check out the code Same wrote here: https://github.com/cbsinteractive/bakery/pull/32/files#diff-f8d027c872dd51f0e133c3c45b300fa9R173-R189 for a way to do this using the `url` package instead of concat-ing strings.

File: pkg/filters/hls.go [hunk 7]
```diff
@@ -200,22 +224,56 @@ func isRelative(urlStr string) (bool, error) {
...
+				if err != nil {
+					return "", fmt.Errorf("trimming segments: %w", err)
+				}
+
+				seqID++
+			}
+		}
+	}
+
+	filteredPlaylist.SeqNo = uint64(seqID)
```
Comment (zsiec) [inline pkg/filters/hls.go]: If we're outputting VOD, shouldn't we keep this at 0?

File: pkg/filters/hls.go (hunk 2 above)
Comment (zsiec) [reply pkg/filters/hls.go]: I see later in a test you've proven it does, ignore me!

File: pkg/filters/hls.go [hunk 8]
```diff
@@ -174,6 +186,18 @@ func (h *HLSFilter) normalizeVariant(v *m3u8.Variant, absolute url.URL) (*m3u8.V
 	return v, nil
 }
 
+func (h *HLSFilter) normalizeTrimmedVariant(filters *parsers.MediaFilters, uri string) (string, error) {
+	encoded := base64.StdEncoding.EncodeToString([]byte(uri))
```
Comment (zsiec) [inline pkg/filters/hls.go]: Because we'll be using this value as a URL path element, I'd recommend we use `RawURLEncoding`: https://golang.org/src/encoding/base64/base64.go?s=6840:6898#L113

If we change this, we have to remember to update the encoding we use when we decode as well.

File: pkg/origin/origin.go [hunk 9]
```diff
@@ -15,46 +15,79 @@ type Origin interface {
 }
 
 //Manifest struct holds Origin and Path of Manifest
+//Variant level manifests will be base64 encoded absolute path
 type Manifest struct {
-	Origin string
-	Path   string
+	Origin   string
+	Path     string
```
Comment (zsiec) [inline pkg/origin/origin.go]: Maybe this should be type `url.URL` so you don't need to cache things like `Absolute` in the struct. Functions like `GetPath` might get simpler as well.

File: pkg/origin/propeller.go [hunk 10]
```diff
@@ -20,28 +22,54 @@ func (p *Propeller) GetPlaybackURL() string {
...
+	case "": //fetch master
+		propellerURL, err = getPropellerChannelURL(c.PropellerHost, orgID, channelID)
+		if err != nil {
+			return &Propeller{}, fmt.Errorf("fetching propeller channel: %w", err)
+		}
+	default:
+		propellerURL, err = decodeRenditionURL(renditionURL)
+		if err != nil {
+			return &Propeller{}, fmt.Errorf("fetching propeller rendition: %w", err)
+		}
```
Comment (zsiec) [inline pkg/origin/propeller.go]: This feels a little awkward to have and empty string for `renditionURL` imply we need to run logic for a master manifest. Perhaps we could try different constructors? It could also be a hint that this struct is doing more than one thing and could use a separation of concerns.

File: pkg/parsers/urlparser.go [hunk 11]
```diff
@@ -54,6 +54,12 @@ const (
 	ProtocolDASH Protocol = "dash"
 )
 
+// Trim is a struct that carries the start and end times to trim playlist
+type Trim struct {
+	Start int `json:",omitempty"`
+	End   int `json:",omitempty"`
```
Comment (zsiec) [inline pkg/parsers/urlparser.go]: Maybe we should explicitly make these `int64` to avoid the casting you do in the filter logic?

File: pkg/parsers/urlparser.go [hunk 12]
```diff
@@ -148,6 +155,17 @@ func URLParse(urlpath string) (string, *MediaFilters, error) {
...
 				mf.MaxBitrate, _ = strconv.Atoi(filters[1])
 			}
+		case "t":
+			var trim Trim
+			if filters[0] != "" {
+				trim.Start, _ = strconv.Atoi(filters[0])
+			}
+
+			if filters[1] != "" {
+				trim.End, _ = strconv.Atoi(filters[1])
```
Comment (zsiec) [inline pkg/parsers/urlparser.go]: Shouldn't we handle these errors to be able to provide useful feedback when we've misconfigured the url?

File: pkg/filters/hls.go [hunk 13]
```diff
@@ -200,22 +224,56 @@ func isRelative(urlStr string) (bool, error) {
...
-		ValidCodecs(codec, ac3Codec))
+// FilterRenditionManifest will be responsible for filtering the manifest
+// according  to the MediaFilters
+func (h *HLSFilter) filterRenditionManifest(filters *parsers.MediaFilters, m *m3u8.MediaPlaylist) (string, error) {
+	filteredPlaylist, _ := m3u8.NewMediaPlaylist(uint(len(m.Segments)), uint(len(m.Segments)))
+	seqID := 0
+
+	for _, segment := range m.Segments {
+		if segment != nil {
+			if inRange(int64(filters.Trim.Start), int64(filters.Trim.End), segment.ProgramDateTime.Unix()) {
```
Comment (zsiec) [inline pkg/filters/hls.go]: Would it make sense to return an error if the value for `segment.ProgramDateTime.Unix()` returned the default value for `time.Time` (signifying the PDT wasn't present)?

File: pkg/origin/origin.go [hunk 14]
```diff
@@ -15,46 +15,79 @@ type Origin interface {
...
+		path = renditionURL
+	}
+
 	return NewManifest(c, path), nil
 }
 
 //NewManifest returns a new Origin struct
 func NewManifest(c config.Config, path string) *Manifest {
+	var absolute bool
+	if strings.Contains(path, "http") {
```
Comment (zsiec) [inline pkg/origin/origin.go]: If we don't decide to change the type for `path` to `url.URL` (which would be nice to be able to use `.IsAbs()`), we should be careful about using `strings.Contains` for checks like this. The string "$$$$$$$http$$$"  would pass this check. We could use `strings.HasPrefix` for some more safety. I think I've done things like `strings.HasPrefix(s, "http://") || strings.HasPrefix(s, "https://")` in the past.

File: pkg/filters/hls.go (hunk 7 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: correct!

File: pkg/filters/hls.go (hunk 5 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: fixed.

File: pkg/filters/hls.go (hunk 4 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: removed this since VOD playlist start at 0

File: pkg/filters/hls.go (hunk 6 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: dope i updated my code a bit to use that. thanks for pointing it out.

File: pkg/parsers/urlparser.go (hunk 11 above)
Comment (krystalmejia24) [reply pkg/parsers/urlparser.go]: done!

File: pkg/parsers/urlparser.go (hunk 12 above)
Comment (krystalmejia24) [reply pkg/parsers/urlparser.go]: good point. i think i got lazy bc i noticed we had no validation in there for bitrate filter conversions. 
Added something small to do both trim and bitrates along with the tests. Simple error check during the conversions from strings and making sure start/min values are not greater than end/max values

File: pkg/origin/propeller.go (hunk 10 above)
Comment (krystalmejia24) [reply pkg/origin/propeller.go]: Im no longer using Propeller struct to manage the encoded rendition urls. This was a first pass and I never cleaned up!

Essentially i move the logic to the Manifest struct. If path is absolute URL that means its the decoded rendition url.

File: pkg/origin/origin.go (hunk 9 above)
Comment (krystalmejia24) [reply pkg/origin/origin.go]: GetPath was in there as a way for me to interface with the propeller struct (which only holds org id and channel id) so that I can get back the original path request `propeller/orgid/channel.m3u8` when writing the rendition urls in the master manifest with the schema `propeller/orgid/channel/base64.m3u8`

I realized when writing those urls that honoring that schema  was overkill since we shouldn't care if the rendition came from propeller or a different origin.

Comment (krystalmejia24) [reply pkg/origin/origin.go]: I've actually updating `Path` to `url.Url` because it reads much cleaner to me and if i need to extent this in the future i think having it as a url type will make it easier to manipulate.

File: pkg/origin/origin.go (hunk 14 above)
Comment (krystalmejia24) [reply pkg/origin/origin.go]: i changed to `url.Url` as described above but I still am using the `path.IsAbs()` bc as you pointed out its better than looking for `http` in the string.

File: pkg/filters/hls.go (hunk 3 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: this one slipped me. good eye.

File: pkg/filters/hls.go (hunk 8 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: updated both encoding and decoding! thanks for this.

File: pkg/filters/hls.go (hunk 13 above)
Comment (krystalmejia24) [reply pkg/filters/hls.go]: done! added a test for this. thanks bc we would have gotten to a situation where we serve an empty playlist for the play to try to serve.

File: pkg/filters/hls.go [hunk 15]
```diff
@@ -200,22 +224,52 @@ func isRelative(urlStr string) (bool, error) {
...
+		return "", fmt.Errorf("filtering Rendition Manifest: %w", err)
+	}
+
+	for _, segment := range m.Segments {
+		if segment == nil {
+			continue
+		}
+
+		if segment.ProgramDateTime == (time.Time{}) {
+			return "", fmt.Errorf("Program Date Time not set on segments")
```
Comment (zsiec) [inline pkg/filters/hls.go]: tinyest of nits:
`Error strings should not be capitalized (unless beginning with proper nouns or acronyms) or end with punctuation, since they are usually printed following other context.`
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: pkg/origin/origin.go [hunk 16]
```diff
@@ -15,9 +17,10 @@ type Origin interface {
 }
 
 //Manifest struct holds Origin and Path of Manifest
+//Variant level manifests will be base64 encoded absolute path
 type Manifest struct {
 	Origin string
-	Path   string
+	Path   url.URL
```
Comment (zsiec) [inline pkg/origin/origin.go]: Maybe we should rename this to `URL` or `OriginalURL`?

Comment (krystalmejia24) [reply pkg/origin/origin.go]: went with URL for clarity
